test: Port `order.rs` tests to sqllogictest

simicd opened the pull request:
## Which issue does this PR close?
Closes #8205

## Rationale for this change
Migrate remaining Rust unit tests in `order.rs` to sqllogictest

## What changes are included in this PR?
- Move unit tests from order.rs to order.slt
- Move parquet file to git submodule

## Are these changes tested?
Yes

## Are there any user-facing changes?
No


-----
Hi everyone, this is my first PR in the arrow-datafusion project (I have previously contributed a few smaller PRs to the arrow-datafusion-python repository). I chose to tackle one of the issues tagged with `good first issue` as I thought it would be a good way to start contributing here. I would greatly appreciate any feedback & suggestions. Thank you in advance for taking the time to review my contribution!

simicd commented on datafusion/core/tests/sql/mod.rs:
```diff
@@ -77,7 +77,6 @@ pub mod csv_files;
 pub mod explain_analyze;
 pub mod expr;
 pub mod joins;
-pub mod order;
```
Removed since there are no more remaining unit tests in `order.rs`

simicd commented on datafusion/core/tests/data/repeat_much.snappy.parquet:
Moved file to same location as other parquet files (git submodule)

simicd commented on datafusion/sqllogictest/test_files/order.slt:
```diff
@@ -578,3 +578,192 @@ SortPreservingMergeExec: [log_c12_base_c11@0 DESC]
+statement ok
+CREATE EXTERNAL TABLE dt (a_id integer, a_str string, a_bool boolean) STORED AS CSV WITH ORDER (a_id ASC) LOCATION 'file://path/to/table';
+
+#TODO: How to check for order in sqllogictest?
```
Here I have a question: How can I check for pre-sorting with pure SQL? 

This is what the Rust unit test would call but I couldn't find anything in the information schema that would return the `file_sort_order` setting.
https://github.com/apache/arrow-datafusion/blob/be361fdd8079a2f44da70f6af6e9d8eb3f7d0020/datafusion/core/tests/sql/order.rs#L60-L62

alamb replied:
One way to test this is to run an explain plan that requires the order. The test should that the data is not re-sorted

For example:

```
# ORDER by query should be able to take advantage of the pre-existing order
# expect to not see SortExec
query EXPLAIN SELECT * FROM dt ORDER BY a_id ASC
```

simicd replied:
Thanks a lot for the fast review and the tip! 

I had to switch the table to an existing query since the old unit test with the inexistent `LOCATION 'file://path/to/table'` would only generate a logical plan, not the physical plan.

So I changed to a TPCH table, without a `WITH ORDER (o_orderkey ASC)` clause the physical plan shows a SortExec:
```
query TT
EXPLAIN SELECT o_orderkey, o_orderstatus FROM orders ORDER BY o_orderkey ASC
----
logical_plan
Sort: orders.o_orderkey ASC NULLS LAST
--TableScan: orders projection=[o_orderkey, o_orderstatus]
physical_plan
SortExec: expr=[o_orderkey@0 ASC NULLS LAST]
--CsvExec: file_groups={1 group: [[WORKSPACE_ROOT/datafusion/core/tests/tpch-csv/orders.csv]]}, projection=[o_orderkey, o_orderstatus], has_header=true
```

Adding it drops it from the physical plan and pushes it into output_ordering in the CsvExec:
```
query TT
EXPLAIN SELECT o_orderkey, o_orderstatus FROM orders ORDER BY o_orderkey ASC
----
logical_plan
Sort: orders.o_orderkey ASC NULLS LAST
--TableScan: orders projection=[o_orderkey, o_orderstatus]
physical_plan CsvExec: file_groups={1 group: [[WORKSPACE_ROOT/datafusion/core/tests/tpch-csv/orders.csv]]}, projection=[o_orderkey, o_orderstatus], output_ordering=[o_orderkey@0 ASC NULLS LAST], has_header=true
```

alamb reviewed (commented):
Thank you @simicd  🙏  -- this looks great. Your contribution is very much apprecaited. 

I left a comment about how to test for ORDER BY in table definition.

simicd commented:
Hi @alamb, thanks a lot for your review and the tip, that worked! I added additional comments above. 

I saw that the previous [CI run failed](https://github.com/apache/arrow-datafusion/actions/runs/7515294363/job/20488918874). I assume this happend because I moved the parquet test file from the main repo into the parquet-testing submodule. Does that mean I have to raise a PR in [this repo first](https://github.com/apache/parquet-testing) and get it merged?

alamb commented on datafusion/sqllogictest/test_files/order.slt:
```diff
@@ -578,3 +578,209 @@ SortPreservingMergeExec: [log_c12_base_c11@0 DESC]
+# Sort with lots of repetition values
+# Test sorting a parquet file with 2 million records that has lots of values that are repeated
+statement ok
+CREATE EXTERNAL TABLE repeat_much STORED AS PARQUET LOCATION '../../parquet-testing/data/repeat_much.snappy.parquet';
```
I think you could use a path like this and leave the file at its current location:

```suggestion
CREATE EXTERNAL TABLE repeat_much STORED AS PARQUET LOCATION '../../datafusion/core/tests/data/repeat_much.snappy.parquet';
```
If it is only used in sqllogictests, maybe you could move the file to 

```
datafusion/sqllogictest/data/repeat_much.snappy.parquet
```

instead

simicd replied:
Great - yes it isn't used anywhere else so I moved it to `datafusion/sqllogictest/data/` as suggested by you

alamb reviewed (commented):
Thanks @simicd  -- this is looking very close -- I think we can avoid having to change parquet-data (which is shared between different parquet implementations) that would be better

alamb reviewed (approved):
Awesome -- thank you so much @simicd
